base/applications/rapps/geninst.cpp

```diff
@@ -0,0 +1,813 @@
+            (tmp = tmp.Mid(0, cch)).AppendFormat(unparamsfmt, L" /S", modechar, bitness, arpkeyname);
+            WriteArpEntry(L"QuietUninstallString", tmp);
+
+            if (GetCustomIconPath(Info, tmp) != S_OK)
+                tmp = Info.MainApp;
+            WriteArpEntry(L"DisplayIcon", tmp);
+
+            if (*GetCommonString(DB_VERSION, tmp))
+                WriteArpEntry(L"DisplayVersion", tmp);
+
+            SYSTEMTIME st;
+            GetSystemTime(&st);
+            tmp.Format(L"%.4u%.2u%.2u", st.wYear, st.wMonth, st.wDay);
+            WriteArpEntry(L"InstallDate", tmp);
+
+            if (*GetCommonString(DB_PUBLISHER, tmp))
+                WriteArpEntry(L"Publisher", tmp);
+
+#if DBG
+            tmp.Format(L"sys64=%d", IsSystem64Bit());
+            WriteArpEntry(L"_DEBUG", tmp);
+#endif
+        }
+
+        if (!Info.Error && Info.ShortcutFile)
+        {
+            CreateShortcut(Info.MainApp);
+        }
+    }
+
+    DeleteDirectoryTree(tempdir.GetString());
+    RemoveDirectory(installdir.GetString()); // This is harmless even if we installed something
+    SendMessage(Info.hDlg, IM_PROGRESS, 0, 0); // Cleanup
+    return 0;
+}
+
+static DWORD CALLBACK
+UIWorkerThread(LPVOID Parameter)
+{
+    ((LPTHREAD_START_ROUTINE)Parameter)(NULL);
+    return SendMessage(g_pInfo->hDlg, IM_END, 0, 0);
+}
+
+static INT_PTR CALLBACK
+UIDlgProc(HWND hDlg, UINT uMsg, WPARAM wParam, LPARAM lParam)
+{
+    HWND hPB = GetDlgItem(hDlg, 1);
+    switch(uMsg)
+    {
+        case IM_STARTPROGRESS:
+            SetWindowLongPtr(hPB, GWL_STYLE, WS_CHILD | WS_VISIBLE);
+            SendMessageW(hPB, PBM_SETMARQUEE, FALSE, 0);
+            SendMessageW(hPB, PBM_SETRANGE32, 0, g_pInfo->Count);
+            SendMessageW(hPB, PBM_SETPOS, 0, 0);
+            break;
+        case IM_PROGRESS:
+            SendMessageW(hPB, PBM_DELTAPOS, 1, 0);
+            break;
+        case IM_END:
+            DestroyWindow(hDlg);
+            break;
+        case WM_INITDIALOG:
+        {
+            SendMessageW(hPB, PBM_SETMARQUEE, TRUE, 0);
+            g_pInfo->hDlg = hDlg;
+            HICON hIco = LoadIconW(hInst, MAKEINTRESOURCEW(IDI_MAIN));
+            SendMessageW(hDlg, WM_SETICON, ICON_BIG, (LPARAM)hIco);
+            SendMessageW(hDlg, WM_SETTEXT, 0, (LPARAM)g_pInfo->AppName);
+            if (!SHCreateThread(UIWorkerThread, (void*)lParam, CTF_COINIT, NULL))
+            {
+                ErrorBox();
+                SendMessageW(hDlg, IM_END, 0, 0);
+            }
+            break;
+        }
+        case WM_CLOSE:
+            return TRUE;
+        case WM_DESTROY:
+            PostMessage(NULL, WM_QUIT, 0, 0);
+            break;
+    }
+    return FALSE;
+}
+
+static BOOL
+CreateUI(BOOL Silent, LPTHREAD_START_ROUTINE ThreadProc)
+{
+    enum { DLGW = 150, DLGH = 20, PAD = 4, PADx2 = PAD * 2, CHILDCOUNT = 1 };
+    const UINT DlgStyle = WS_CAPTION | DS_MODALFRAME | DS_NOFAILCREATE | DS_CENTER;
+    static const WORD DlgTmpl[] =
```

Comment:
Why is this template defined inline, and not in the resources?

Reply:
There is nothing to translate nor edit. If we end up wanting a real UI with checkbox for start menu shortcut or something like that it should become a resource.